apollo-router/src/test_harness/http_snapshot.rs

```diff
@@ -0,0 +1,520 @@
+            let name = name.to_string();
+            if include(&name) {
+                let value = value.to_str().unwrap_or_default().to_string();
+                map.entry(name).or_default().push(value);
+            }
+            map
+        },
+    )
+}
+
+fn save<P: AsRef<Path>>(path: P, snapshots: &mut Vec<Snapshot>) -> Result<(), SnapshotError> {
+    let path = path.as_ref();
+    if let Some(parent) = path.parent() {
+        std::fs::create_dir_all(parent)?;
+    }
+
+    // Keep the snapshots sorted to make source control merges easier when updating snapshots
+    snapshots.sort_by_key(|snapshot| {
+        format!(
+            "{}-{}",
+            &snapshot
+                .request
+                .method
+                .as_ref()
+                .unwrap_or(&String::from("GET")),
+            &snapshot.request.path.as_ref().unwrap_or(&String::from("/"))
+        )
+    });
+
+    std::fs::write(path, serde_json::to_string_pretty(&snapshots)?).map_err(Into::into)
+}
+
+fn load<P: AsRef<Path>>(path: P) -> Result<Vec<Snapshot>, SnapshotError> {
+    let str = std::fs::read_to_string(path)?;
+    let result: Result<Vec<Snapshot>, SnapshotError> =
+        serde_json::from_str(&str).map_err(Into::into);
+    if let Ok(snapshots) = &result {
+        info!("Loaded {} snapshots", snapshots.len());
+    }
+    result
+}
+
+impl SnapshotServer {
+    /// Spawn the server in a new task and return. Used for tests.
+    #[cfg_attr(test, allow(unreachable_pub))]
+    pub async fn spawn<P: AsRef<Path>>(
+        snapshot_path: P,
+        base_url: Uri,
+        offline: bool,
+        update: bool,
+        include_headers: Option<Vec<String>>,
+    ) -> Self {
+        Self::inner_start(
+            snapshot_path,
+            base_url,
+            true,
+            offline,
+            update,
+            include_headers,
+            None,
+        )
+        .await
+    }
+
+    /// Start the server and block. Can be used to run the server as a standalone application.
+    pub(crate) async fn start<P: AsRef<Path>>(
+        snapshot_path: P,
+        base_url: Uri,
+        offline: bool,
+        update: bool,
+        include_headers: Option<Vec<String>>,
+        listener: Option<TcpListener>,
+    ) -> Self {
+        Self::inner_start(
+            snapshot_path,
+            base_url,
+            false,
+            offline,
+            update,
+            include_headers,
+            listener,
+        )
+        .await
+    }
+
+    /// Get the URI the server is listening at
+    #[cfg_attr(not(test), allow(dead_code))]
+    #[cfg_attr(test, allow(unreachable_pub))]
+    pub fn uri(&self) -> String {
+        format!("http://{}", self.socket_address)
+    }
+
+    async fn inner_start<P: AsRef<Path>>(
+        snapshot_path: P,
+        base_url: Uri,
+        spawn: bool,
+        offline: bool,
+        update: bool,
+        include_headers: Option<Vec<String>>,
+        listener: Option<TcpListener>,
```

Comment:
Might be useful to encode `spawn` and `listener` into an enum, since they (seem) to always need to be the two following values:

- `spawn: true` and `listener: None`
- `spawn: false` and `listener: Some(...)`

Reply:
They're unrelated concerns, it's just used that way today.